### DISCUSSION TITLE
Questions about environment variables

### DISCUSSION THREAD

**Discussion by helmut72** [zitadel/zitadel#4596] [Q&A]
Hi!

Since it looks like that LDAP support for the backend is coming into Zitadel, I thought to try Zitadel if it can replace Keycloak or Authelia for webapps, that supports OIDC or SAML. In a homelab. A few users, all managed in LDAP. Not all apps supports OIDC or SAML.

But I feel I'm far away to understand even Zitadels setup ...

1. DEFAULTINSTANCE or FIRSTINSTANCE?

This says defaultinstance:
https://github.com/zitadel/zitadel/blob/c2a5b785fb627a92d4bde6cb44fb8bdfe11ee116/cmd/defaults.yaml#L317

... with description from this link:
https://docs.zitadel.com/docs/guides/manage/self-hosted/configure#environment-variables

```
Prefix the key with ZITADEL_, uppercase the propery and join sections by an underscore _.
```

For me this means I should use `ZITADEL_DEFAULTINSTANCE_...` because it's `DefaultInstance:`. But then:

```
For example, if you want to configure the default ZITADEL IAM admin username and password, make sure the ZITADEL binary runtime has the variables ZITADEL_FIRSTINSTANCE_ORG_HUMAN_USERNAME
```

So I tried FIRSTINSTANCE and had more luck, but what is the right name? Also FIRSTINSTANCE doesn't work for all settings ...

2. Some ENV variables are not set

Example 1:

```
ZITADEL_FIRSTINSTANCE_ORG_HUMAN_PASSWORD=example1234

ZITADEL_FIRSTINSTANCE_PASSWORDCOMPLEXITYPOLICY_MINLENGTH=8
ZITADEL_FIRSTINSTANCE_PASSWORDCOMPLEXITYPOLICY_HASLOWERCASE=true
ZITADEL_FIRSTINSTANCE_PASSWORDCOMPLEXITYPOLICY_HASUPPERCASE=false
ZITADEL_FIRSTINSTANCE_PASSWORDCOMPLEXITYPOLICY_HASNUMBER=true
ZITADEL_FIRSTINSTANCE_PASSWORDCOMPLEXITYPOLICY_HASSYMBOL=false
```

This password doesn't work even if complexity matches.

Example 2:

```
ZITADEL_FIRSTINSTANCE_SMTPCONFIGURATION_SMTP_HOST=mail.example.com:587
ZITADEL_FIRSTINSTANCE_SMTPCONFIGURATION_SMTP_USER=serviceuser
ZITADEL_FIRSTINSTANCE_SMTPCONFIGURATION_SMTP_PASSWORD=servicepassword
ZITADEL_FIRSTINSTANCE_SMTPCONFIGURATION_FROM=zitadel@example.com
ZITADEL_FIRSTINSTANCE_SMTPCONFIGURATION_FROMNAME=Zitadel
ZITADEL_FIRSTINSTANCE_SMTPCONFIGURATION_TLS=true
```

SMTP settings are still unset in the webinterface. Also some others settings are not changed in the webinterface.

What is my mistake for both examples?

3. Why can't I add admins real mail address with ENV variables?

This is required to access Zitadel when I run it behind my reverse proxy:

```
ZITADEL_EXTERNALDOMAIN=zitadel.example.com
```

Also I think this is required:

```
ZITADEL_FIRSTINSTANCE_ORG_NAME=MYORG
```

I use this ENV variable for the user:

```
ZITADEL_FIRSTINSTANCE_ORG_HUMAN_USERNAME=myadmin
```

Zitadel generates a login name like this: `myadmin@myorg.zitadel.example.com`

In the webinterface I see a generated mail address: `admin@zitadel.example.com`

But I don't use a subdomain for my mail domain, just a domain. Like nearly everyone. Also the name `admin` is something I don't have set anywhere. When I set `EXTERNALDOMAIN` to `example.com`, of course Zitadel isn't accessable anymore with `zitadel.example.com`.

Then I've tried this one:

```
ZITADEL_FIRSTINSTANCE_ORG_HUMAN_EMAIL_ADDRESS=myadmin@example.com
ZITADEL_FIRSTINSTANCE_ORG_HUMAN_EMAIL_VERIFIED=true
```

But running with this setting only rise the highscore for "how many clean starts do I need to setup a new app with the right env variables'. Means I can't login with `myadmin@myorg.zitadel.example.com` anymore.

What is my mistake here?

4. mail address II / usernames / certificates

Nothing regarding environment variables, but it's follow up. In the example above the admin got the mail address `admin@zitadel.example.com` and also something like a mail address: `myadmin@myorg.zitadel.example.com`. Is the latter used to receive mails?
Is it possible that usernames can login with their username instead of their mail address? It's just one/my domain.

Also in the webinterface (instance menu) there is something like `zitadel-abc123.zitadel.example.com`. Do I need a (wildcard?) certificate for subdomain `zitadel.example.com` too or is `zitadel-abc123...` unused from any app or browser?

Ok, I come to the end ... ;-) 

I'm using `:stable` from yesterday and all I want to do is prefill my prefered settings to avoid clicking through all menues.

Thank you.

**Comment by fforootd** [comment]
Hi @helmut72 thanks for this nice write down. 

Let us come back with an in depth answer ASAP 😁 

I think some of the confusion is down to some outdated stuff we need to clean (Especially the FIRST and DEFAULT part)

CC @zitadel/engineers

**Comment by helmut72** [comment]
Thank you 😊

In the meantime I've stopped trying setup Zitadel with ENV variables and played today with Zitadel.

> Also in the webinterface (instance menu) there is something like `zitadel-abc123.zitadel.example.com`. Do I need a (wildcard?) certificate for subdomain `zitadel.example.com` too or is `zitadel-abc123...` unused from any app or browser?

It looks like an instance is the installation itself. It feels there can be more instances running and I don't need to care about this names, which looks like a subdomain. I found also `orga1.zitadel.example.com`, `orga2.zitadel.example.com` .. in the organization menu. Both, the "subdomains" in instances and organizations, looks like they are just for internal usage. Am I right? My guess: I have only accessed `zitadel.example.com` from my Browser (at least for now 😉).

> Is it possible that usernames can login with their username instead of their mail address? It's just one/my domain.

Looks like it's possible: `Settings -> Domain settings`. My admin user and my test user can now login with their username.

> In the example above the admin got the mail address `admin@zitadel.example.com` and also something like a mail address: `myadmin@myorg.zitadel.example.com`. Is the latter used to receive mails?

It looks like `myadmin@myorg.zitadel.example.com` is for internal usage and it's a (long) username, when the setting `User Loginname must contain orgdomain` is enabled.
Mails will be send only to `admin@zitadel.example.com`. If I don't change it. Because I have renamed it to a vaild mail address and received a verification mail. Am I right? Then this question is answered.

Then I have another question. When `User Loginname must contain orgdomain` is disabled, I can't create a new user `user1` with the same username that exists in a different organization. Also `user1`.
But if I enable `User Loginname must contain orgdomain` it's possible to create a new user `user1` with the same username that exists in a different organization. When disabling `User Loginname must contain orgdomain` again on instance and organizations settings, which user will be then logged in? `user1` in organization `orga1` or `user1` in organization `orga2`?

Is it possible that a user can also login with one of his mail addresses? His username and mail addresses are more common for casual users than any `user1@orga.zitadel.example.com`.

**Comment by adlerhurst** [comment]
Hi @helmut72 

Thanks for your interest in ZITADEL ☺️

If I miss a question please let me know.

> DEFAULTINSTANCE or FIRSTINSTANCE?

Instances are used to run multiple installations in a single binary. This is used for cases like zitadel.cloud. Most local use cases will only require one instance.

`DEFAULTINSTANCE`: If a new instance is created, the initial data of an instance have to be stored. These data are read from this configuration.

`FIRSTINSTANCE`: During the setup of ZITADEL an instance has to be created because you cannot interact with ZITADEL without an instance. The first instance allows to overwrite some parts of the `DEFAULTINSTANCE`-config but only for the instance created during setup and only once. 

You can set the following fields:
- `InstanceName`
- `DefaultLanguage`
- `Org`
  - `Name`
  - `Human`
    - ...

> Some ENV variables are not set

The password complexity policy and smtp configuration are only possible to set on `DEFAULTINSTANCE`. 

> Why can't I add admins real mail address with ENV variables?

using the env variable you mentioned it should use the provided email address

> This is required to access Zitadel when I run it behind my reverse proxy:

Did you start zitadel with the correct tls-mode? https://docs.zitadel.com/docs/guides/manage/self-hosted/tls_modes

An example on how to configure zitadel behind a load balancer can be found here: https://docs.zitadel.com/docs/guides/deploy/loadbalancing-example

> When disabling User Loginname must contain orgdomain again on...

Here is the tracking issue for the problem: https://github.com/zitadel/zitadel/issues/4487

**Comment by helmut72** [comment]
@adlerhurst 

Thanks for the answer!

> The password complexity policy and smtp configuration are only possible to set on DEFAULTINSTANCE.

I have also tried a mix of DEFAULTINSTANCE and FIRSTINSTANCE, but maybe the opposite. Anyway, it works now, with DEFAULTINSTANCE for smtp, password complexity and FIRSTINSTANCE for all user related information.

> Did you start zitadel with the correct tls-mode? https://docs.zitadel.com/docs/guides/manage/self-hosted/tls_modes

TLS is no problem if Zitadel works like Authelia or Keycloak. Both can be configured and accessed with something like auth.example.com. Because I'm used to this I have configured Zitadel to access it with zitadel.example.com. I don't want use just the domain name (example.com) to access Zitadel.

I have more a question if I also need a certificate for the subdomain zitadel.example.com? From the instance view, there is a "hostname" (zitadel-k8yajt) in the subdomain zitadel.example.com:

![zitadel1](https://user-images.githubusercontent.com/1388751/197688195-bd46af3c-4cb4-40e9-b999-927ff769baed.jpg)

... and also for the organization, because in the organization view there is something like `orga1.zitadel.example.com` when setup the required option `ZITADEL_FIRSTINSTANCE_ORG_NAME`?

Will both FQDNS (zitadel-k8yajt.zitadel.example.com and orga1.zitadel.example.com) accessed from a browser (or an app) or are both only used for Zitadel-internal usage?

My other questions (for now) will be answered if following issue is closed:
https://github.com/zitadel/zitadel/issues/3976

Thank you.

**Comment by helmut72** [comment]
> Did you start zitadel with the correct tls-mode? https://docs.zitadel.com/docs/guides/manage/self-hosted/tls_modes

To explain it with configuration files. With following I can access Zitadel with https://zitadel.example.com and the primary organization domain becomes `orga1.zitadel.example.com`:

```
ZITADEL_EXTERNALDOMAIN=zitadel.example.com
ZITADEL_FIRSTINSTANCE_ORG_NAME=ORGA1

ZITADEL_EXTERNALPORT=443
ZITADEL_EXTERNALSECURE=true
ZITADEL_TLS_ENABLED=false
```

With following I **can't** access Zitadel with https://zitadel.example.com anymore and I think, the primary organization domain becomes `orga1.example.com`.

```
ZITADEL_EXTERNALDOMAIN=example.com
ZITADEL_FIRSTINSTANCE_ORG_NAME=ORGA1

ZITADEL_EXTERNALPORT=443
ZITADEL_EXTERNALSECURE=true
ZITADEL_TLS_ENABLED=false
```

Both are not my primary domain. It's example.com, but https://example.com is common to use it for a websites nowadays.

Regarding `orga1.zitadel.example.com` I have asked if `orga1.zitadel.example.com` will be access from any app or browser? My certificate is for example.com, not for the subdomain zitadel.example.com
